File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/BlockHeaderPacket.java
package net.ballmerlabs.uscatterbrain.network;

import android.util.Log;

import net.ballmerlabs.uscatterbrain.ScatterProto;
import net.ballmerlabs.uscatterbrain.db.ScatterbrainDatastore;
import com.github.davidmoten.rx2.Bytes;
import com.google.protobuf.ByteString;
import com.google.protobuf.GeneratedMessageLite;
import com.goterl.lazycode.lazysodium.interfaces.Sign;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.List;
import java.util.UUID;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * Wrapper class for protocol buffer blockdata message
 */
public class BlockHeaderPacket implements ScatterSerializable {
    private ScatterProto.BlockData blockdata;
    private final List<ByteString> mHashList;
    private final ByteString mFromFingerprint;
    private final ByteString mToFingerprint;
    private final String extension;
    private byte[] mSignature;
    private final byte[] mApplication;
    private final int mSessionID;
    private boolean mToDisk;
    private boolean endofstream;
    private final int mBlocksize;
    private final String mime;
    private final String filename;
    private UUID luidtag;

    private BlockHeaderPacket(Builder builder) {
        this.mHashList = builder.getHashlist();
        this.endofstream = builder.endofstream;
        this.extension = builder.extension;
        if (builder.getSig() == null) {
            this.mSignature = new byte[Sign.ED25519_BYTES];
        } else {
            this.mSignature = builder.getSig();
        }
        if (builder.getmToFingerprint() != null) {
            this.mToFingerprint = ByteString.copyFrom(builder.getmToFingerprint());
        } else {
            this.mToFingerprint = ByteString.EMPTY;
        }

        if (builder.getmFromFingerprint() != null) {
            this.mFromFingerprint = ByteString.copyFrom(builder.getmFromFingerprint());
        } else  {
            this.mFromFingerprint = ByteString.EMPTY;
        }
        this.mApplication = builder.getApplication();
        this.mSessionID = builder.getSessionid();
        this.mToDisk = builder.getToDisk();
        this.mBlocksize = builder.getBlockSize();
        this.mime = builder.mime;
        final ScatterProto.BlockData.Builder b = ScatterProto.BlockData.newBuilder();
        if (builder.filename == null) {
            this.filename = getAutogenFilename();
            b.setFilenameGone(true);
        } else {
            b.setFilenameVal(builder.filename);
            this.filename = builder.filename;
        }
        regenBlockData();
    }

    public void markEnd() {
        this.endofstream = true;
        regenBlockData();
    }

    public boolean isEndOfStream() {
        return this.endofstream;
    }

    private ByteString sumBytes() {
        ByteString messagebytes = ByteString.EMPTY;
        messagebytes = messagebytes.concat(this.mFromFingerprint);
        messagebytes = messagebytes.concat(this.mToFingerprint);
        messagebytes = messagebytes.concat(ByteString.copyFrom(this.mApplication));
        messagebytes = messagebytes.concat(ByteString.copyFromUtf8(this.extension));
        messagebytes = messagebytes.concat(ByteString.copyFromUtf8(this.mime));
        messagebytes = messagebytes.concat(ByteString.copyFromUtf8(this.filename));
        byte td = 0;
        if (this.mToDisk)
            td = 1;
        ByteString toDiskBytes = ByteString.copyFrom(ByteBuffer.allocate(1).order(ByteOrder.BIG_ENDIAN).put(td).array());
        messagebytes = messagebytes.concat(toDiskBytes);

        for (ByteString hash : this.mHashList) {
            messagebytes = messagebytes.concat(hash);
        }
        return messagebytes;
    }

    /**
     * Verifyed 25519 boolean.
     *
     * @param pubkey the pubkey
     * @return the boolean
     */
    public boolean verifyed25519(byte[] pubkey) {
        if (pubkey.length != Sign.PUBLICKEYBYTES)
            return false;

        ByteString messagebytes = sumBytes();
        return LibsodiumInterface.getSodium().crypto_sign_verify_detached(this.blockdata.getSig().toByteArray(),
                messagebytes.toByteArray(),
                messagebytes.size(),
                pubkey) == 0;
    }

    private void regenBlockData() {
        this.blockdata = ScatterProto.BlockData.newBuilder()
                .setApplicationBytes(ByteString.copyFrom(this.mApplication))
                .setFromFingerprint(this.mFromFingerprint)
                .setToFingerprint(this.mToFingerprint)
                .setTodisk(this.mToDisk)
                .setExtension(this.extension)
                .addAllNexthashes(this.mHashList)
                .setSessionid(this.mSessionID)
                .setBlocksize(this.mBlocksize)
                .setMime(this.mime)
                .setEndofstream(this.endofstream)
                .setSig(ByteString.copyFrom(this.mSignature))
                .build();
    }

    /**
     * Sign ed 25519 boolean.
     *
     * @param secretkey the secretkey
     * @return the boolean
     */
    public boolean signEd25519(byte[] secretkey) {
        if (secretkey.length != Sign.SECRETKEYBYTES)
            return false;

        ByteString messagebytes = sumBytes();

        this.mSignature = new byte[Sign.ED25519_BYTES];
        Pointer p = new PointerByReference(Pointer.NULL).getPointer();
        if (LibsodiumInterface.getSodium().crypto_sign_detached(this.mSignature,
                p, messagebytes.toByteArray(), messagebytes.size(), secretkey) == 0) {
            regenBlockData();
            return true;
        } else {
            return false;
        }
    }

    public String getUserFilename() {
        return filename;
    }

    public String getAutogenFilename() {
        String ext = ScatterbrainDatastore.getDefaultFileName(this) + "." +
                ScatterbrainDatastore.sanitizeFilename(extension);
        Log.e("debug", "getAutogenFilename: " + ext);
        return ext;
    }

    private BlockHeaderPacket(InputStream in) throws IOException {
        this.blockdata = CRCProtobuf.parseFromCRC(ScatterProto.BlockData.parser(), in);
        this.mApplication = blockdata.getApplicationBytes().toByteArray();
        Log.e("debug ", "header nexthashes count" + blockdata.getNexthashesList().size());
        Log.e("debug", "header nexthashes raw count " + blockdata.getNexthashesCount());
        this.mHashList = blockdata.getNexthashesList();
        this.mFromFingerprint = blockdata.getFromFingerprint();
        this.mToFingerprint = blockdata.getToFingerprint();
        this.mSignature = blockdata.getSig().toByteArray();
        this.mToDisk = blockdata.getTodisk();
        this.mSessionID = blockdata.getSessionid();
        this.mBlocksize = blockdata.getBlocksize();
        this.extension = blockdata.getExtension();
        this.endofstream = blockdata.getEndofstream();
        if (blockdata.getFilenameCase().equals(ScatterProto.BlockData.FilenameCase.FILENAME_VAL)) {
            this.filename = blockdata.getFilenameVal();
        } else {
            this.filename = getAutogenFilename();
        }
        this.mime = blockdata.getMime();
    }

    /**
     * Parse from blockheader packet.
     *
     * @param is the is
     * @return the block header packet
     */
    public static Single<BlockHeaderPacket> parseFrom(InputStream is) {
        return Single.fromCallable(() -> new BlockHeaderPacket(is));
    }

    public static Single<BlockHeaderPacket> parseFrom(Observable<byte[]> flowable) {
        InputStreamObserver observer = new InputStreamObserver();
        flowable.subscribe(observer);
        return BlockHeaderPacket.parseFrom(observer).doFinally(observer::close);
    }

    public static Single<BlockHeaderPacket> parseFrom(Flowable<byte[]> flowable) {
        InputStreamFlowableSubscriber observer = new InputStreamFlowableSubscriber();
        flowable.subscribe(observer);
        return parseFrom(observer).doFinally(observer::close);
    }

    @Override
    public void tagLuid(UUID luid) {
        luidtag = luid;
    }

    @Override
    public UUID getLuid() {
        return luidtag;
    }

    @Override
    public byte[] getBytes() {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        try {
            CRCProtobuf.writeToCRC(blockdata, os);
            return os.toByteArray();
        } catch (IOException e) {
            return null;
        }
    }

    @Override
    public ByteString getByteString() {
        return ByteString.copyFrom(getBytes());
    }

    @Override
    public Completable writeToStream(OutputStream os) {
        return Completable.fromAction(() -> CRCProtobuf.writeToCRC(blockdata, os));
    }

    @Override
    public Flowable<byte[]> writeToStream(int fragsize) {
        return Bytes.from(new ByteArrayInputStream(getBytes()), fragsize);
    }

    @Override
    public GeneratedMessageLite getMessage() {
        return blockdata;
    }

    @Override
    public PacketType getType() {
        return PacketType.TYPE_BLOCKHEADER;
    }

    /**
     * Gets blockdata.
     *
     * @return the blockdata
     */
    public ScatterProto.BlockData getBlockdata() {
        return this.blockdata;
    }

    /**
     * Gets the blocksize
     * @return int blocksize
     */
    public int getBlockSize() {
        return this.blockdata.getBlocksize();
    }

    /**
     * Gets hash.
     *
     * @param seqnum the seqnum
     * @return the hash
     */
    public ByteString getHash(int seqnum) {
        return this.blockdata.getNexthashes(seqnum);
    }

    /**
     * Gets sig.
     *
     * @return the sig
     */
    public ByteString getSig() {
        return ByteString.copyFrom(this.mSignature);
    }

    /**
     * Get application byte [ ].
     *
     * @return the byte [ ]
     */
    public byte[] getApplication() {
        return this.mApplication;
    }

    /**
     * Gets hash list.
     *
     * @return the hash list
     */
    public List<ByteString> getHashList() {
        return mHashList;
    }

    /**
     * Gets from fingerprint.
     *
     * @return the from fingerprint
     */
    public ByteString getFromFingerprint() {
        return mFromFingerprint;
    }

    /**
     * Gets to fingerprint.
     *
     * @return the to fingerprint
     */
    public ByteString getToFingerprint() {
        return mToFingerprint;
    }

    /**
     * Get signature byte [ ].
     *
     * @return the byte [ ]
     */
    public byte[] getSignature() {
        return mSignature;
    }

    /**
     * Gets session id.
     *
     * @return the session id
     */
    public int getSessionID() {
        return mSessionID;
    }

    /**
     * Gets to disk.
     *
     * @return the to disk
     */
    public boolean getToDisk() {
        return mToDisk;
    }

    /**
     * Gets file extension
     * @return file extension
     */
    public String getExtension() {
        return ScatterbrainDatastore.sanitizeFilename(extension);
    }

    /**
     * Sets to disk.
     *
     * @param t the t
     */
    public void setToDisk(boolean t) { this.mToDisk = t; }

    public String getMime() {
        return mime;
    }

    /**
     * New builder builder.
     *
     * @return the builder
     */
    public static Builder newBuilder() {
        return new Builder();
    }

    /**
     * The type Builder.
     */
    public static class Builder {
        private boolean todisk;
        private byte[] application;
        private int sessionid;
        public int mBlockSize;
        private byte[] mToFingerprint;
        private byte[] mFromFingerprint;
        private String extension;
        private List<ByteString> hashlist;
        private byte[] mSig;
        private String filename;
        private String mime;
        private boolean endofstream = false;

        /**
         * Instantiates a new Builder.
         */
        public Builder() {
            todisk = false;
            sessionid = -1;
            mBlockSize = -1;
            mime = "application/octet-stream";
        }

        /**
         * Sets the fingerprint for the recipient.
         *
         * @param toFingerprint the to fingerprint
         * @return builder
         */
        public Builder setToFingerprint(byte[] toFingerprint) {
            this.mToFingerprint = toFingerprint;
            return this;
        }

        /**
         * Sets from fingerprint.
         *
         * @param fromFingerprint sets the fingerprint for the sender
         * @return builder
         */
        public Builder setFromFingerprint(byte[] fromFingerprint) {
            this.mFromFingerprint = fromFingerprint;
            return this;
        }

        /**
         * Sets application.
         *
         * @param application bytes for UTF encoded scatterbrain application string
         * @return builder
         */
        public Builder setApplication(byte[] application) {
            this.application = application;
            return this;
        }

        /**
         * Sets to disk.
         *
         * @param toDisk whether to write this file to disk or attempt to store it in the database
         * @return builder
         */
        public Builder setToDisk(boolean toDisk) {
            this.todisk = toDisk;
            return this;
        }

        /**
         * Sets session id.
         *
         * @param sessionID the session id (used for upgrading between protocols)
         * @return builder
         */
        public Builder setSessionID(int sessionID) {
            this.sessionid = sessionID;
            return this;
        }

        /**
         * Sets hashes.
         *
         * @param hashes list of hashes of following blocksequence packets.
         * @return builder
         */
        public Builder setHashes(List<ByteString> hashes) {
            this.hashlist = hashes;
            return this;
        }


        /**
         * Sets the file extension
         * @param ext: string file extension
         * @return builder
         */
        public Builder setExtension(String ext) {
            this.extension = ext;
            return this;
        }

        /**
         * Sets blocksize
         * @param blockSize
         * @return builder
         */
        public Builder setBlockSize(int blockSize) {
            this.mBlockSize = blockSize;
            return this;
        }

        public Builder setSig(byte[] sig) {
            this.mSig = sig;
            return this;
        }

        public Builder setMime(String mime) {
            this.mime = mime;
            return this;
        }

        public Builder setEndOfStream() {
            this.endofstream = true;
            return this;
        }

        public Builder setFilename(String filename) {
            this.filename = filename;
            return this;
        }

        public byte[] getSig() {
            return mSig;
        }

        /**
         * Build block header packet.
         *
         * @return the block header packet
         */
        public BlockHeaderPacket build() {
            if (hashlist == null)
                throw new IllegalArgumentException("hashlist was null");

            // fingerprints and application are required
            if (application == null) {
                throw new IllegalArgumentException("application was null");
            }

            if (extension == null) {
                extension = ".data";
            }

            if (mBlockSize <= 0) {
                IllegalArgumentException e = new IllegalArgumentException("blocksize not set");
                e.printStackTrace();
                throw e;
            }

            return new BlockHeaderPacket(this);
        }

        /**
         * Get application name (in UTF-8).
         *
         * @return the byte [ ]
         */
        public byte[] getApplication() {
            return application;
        }

        /**
         * Gets sessionid.
         *
         * @return the sessionid
         */
        public int getSessionid() {
            return sessionid;
        }

        /**
         * Gets hashlist.
         *
         * @return the hashlist
         */
        public List<ByteString> getHashlist() {
            return hashlist;
        }

        /**
         * Gets to fingerprint.
         *
         * @return the to fingerprint
         */
        public byte[] getmToFingerprint() {
            return mToFingerprint;
        }

        /**
         * Gets from fingerprint.
         *
         * @return the from fingerprint
         */
        public byte[] getmFromFingerprint() {
            return mFromFingerprint;
        }

        /**
         * Gets to disk.
         *
         * @return the to disk
         */
        public boolean getToDisk() {
            return todisk;
        }

        /**
         * Gets blocksize
         * @return blocksize
         */
        public int getBlockSize() { return mBlockSize; }
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/BlockHeaderPacket.kt
package net.ballmerlabs.uscatterbrain.network

import android.util.Log
import com.google.protobuf.ByteString
import net.ballmerlabs.uscatterbrain.ScatterProto
import net.ballmerlabs.uscatterbrain.ScatterProto.BlockData
import net.ballmerlabs.uscatterbrain.db.getDefaultFileName
import net.ballmerlabs.uscatterbrain.db.sanitizeFilename

/**
 * Wrapper class for protocol buffer blockdata message
 */
class BlockHeaderPacket(blockdata: BlockData) : ScatterSerializable<BlockData>(blockdata), Verifiable  {
    /**
     * Gets hash list.
     *
     * @return the hash list
     */
    val hashList
    get() = packet.nexthashesList

    override val hashes
        get() = hashList.map { v -> v.toByteArray() }.toTypedArray()


    /**
     * Gets from fingerprint.
     *
     * @return the from fingerprint
     */
    override val fromFingerprint: ByteArray?
        get() {
            val r = packet.fromFingerprint.toByteArray()
            return if (r.size == 1) null else r
        }


    /**
     * Gets to fingerprint.
     *
     * @return the to fingerprint
     */
    override val toFingerprint: ByteArray?
        get() {
            val r = packet.toFingerprint.toByteArray()
            return if (r.size == 1) null else r
        }

    override val extension: String
        get() = sanitizeFilename(packet.extension)

    /**
     * Get signature byte [ ].
     *
     * @return the byte [ ]
     */
    override val signature: ByteArray?
        get() = if (fromFingerprint == null) null else packet.sig.toByteArray()

    /**
     * Get application byte [ ].
     *
     * @return the byte [ ]
     */
    override val application: String
        get() = packet.application

    /**
     * Gets session id.
     *
     * @return the session id
     */
    val sessionID: Int
        get() = packet.sessionid

    override val toDisk: Boolean
        get() = packet.todisk

    val isEndOfStream: Boolean
        get() = packet.endofstream

    private val mBlocksize: Int
        get() = packet.blocksize

    override val mime: String
        get() = packet.mime

    override val userFilename: String
    get() = packet.filename

    val autogenFilename: String
        get() {
            if (isEndOfStream) {
                return ""
            }

            val ext: String = getDefaultFileName(this) + "." + extension
            Log.e("debug", "getAutogenFilename: $ext")
            return ext
        }

    override val type: PacketType
        get() = PacketType.TYPE_BLOCKHEADER

    /**
     * Gets the blocksize
     * @return int blocksize
     */
    val blockSize: Int
        get() = packet.blocksize

    /**
     * Gets hash.
     *
     * @param seqnum the seqnum
     * @return the hash
     */
    fun getHash(seqnum: Int): ByteString {
        return packet.getNexthashes(seqnum)
    }

    /**
     * The type Builder.
     */
    data class Builder(
            var toDisk: Boolean = false,
            var application: ByteArray? = null,
            var sessionid: Int? = null,
            var blockSizeVal: Int? = null,
            var mToFingerprint: ByteArray? = null,
            var mFromFingerprint: ByteArray? = null,
            var extensionVal: String = "",
            var hashlist: List<ByteString>? = null,
            var sig: ByteArray? = null,
            var filename: String? = null,
            var mime: String? = null,
            var endofstream: Boolean = false,

    ) {

        /**
         * Sets the fingerprint for the recipient.
         *
         * @param toFingerprint the to fingerprint
         * @return builder
         */
        fun setToFingerprint(toFingerprint: ByteArray?) = apply {
            mToFingerprint = toFingerprint
        }

        /**
         * Sets from fingerprint.
         *
         * @param fromFingerprint sets the fingerprint for the sender
         * @return builder
         */
        fun setFromFingerprint(fromFingerprint: ByteArray?) = apply {
            mFromFingerprint = fromFingerprint
        }

        /**
         * Sets application.
         *
         * @param application bytes for UTF encoded scatterbrain application string
         * @return builder
         */
        fun setApplication(application: ByteArray) = apply {
            this.application = application
        }
        /**
         * Sets to disk.
         *
         * @param toDisk whether to write this file to disk or attempt to store it in the database
         * @return builder
         */
        fun setToDisk(toDisk: Boolean) = apply {
            this.toDisk = toDisk
        }

        /**
         * Sets session id.
         *
         * @param sessionID the session id (used for upgrading between protocols)
         * @return builder
         */
        fun setSessionID(sessionID: Int) = apply {
            sessionid = sessionID
        }

        /**
         * Sets hashes.
         *
         * @param hashes list of hashes of following blocksequence packets.
         * @return builder
         */
        fun setHashes(hashes: List<ByteString>) = apply {
            hashlist = hashes
        }

        /**
         * Sets the file extension
         * @param ext: string file extension
         * @return builder
         */
        fun setExtension(ext: String) = apply {
            this.extensionVal = ext
        }

        /**
         * Sets blocksize
         * @param blockSize
         * @return builder
         */
        fun setBlockSize(blockSize: Int) = apply {
            this.blockSizeVal = blockSize
        }

        fun setSig(sig: ByteArray?) = apply {
            this.sig = sig
        }

        fun setMime(mime: String) = apply {
            this.mime = mime
        }

        fun setEndOfStream(value: Boolean) = apply {
            endofstream = value
        }

        fun setFilename(filename: String?) = apply {
            this.filename = filename
        }

        /**
         * Build block header packet.
         *
         * @return the block header packet
         */
        fun build(): BlockHeaderPacket {
            if (!endofstream) {
                if (blockSizeVal!! <= 0) {
                    val e = IllegalArgumentException("blocksize not set")
                    e.printStackTrace()
                    throw e
                }
            }
            val packet = BlockData.newBuilder()
                    .setApplicationBytes(ByteString.copyFrom(application!!))
                    .setFromFingerprint(ByteString.copyFrom(mFromFingerprint))
                    .setToFingerprint(ByteString.copyFrom(mToFingerprint))
                    .setTodisk(toDisk)
                    .setExtension(extensionVal)
                    .addAllNexthashes(hashlist)
                    .setSessionid(sessionid!!)
                    .setBlocksize(blockSizeVal!!)
                    .setMime(mime)
                    .setEndofstream(endofstream)
                    .setSig(ByteString.copyFrom(sig))
                    .build()

            return BlockHeaderPacket(packet)
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Builder

            if (toDisk != other.toDisk) return false
            if (application != null) {
                if (other.application == null) return false
                if (!application.contentEquals(other.application)) return false
            } else if (other.application != null) return false
            if (sessionid != other.sessionid) return false
            if (blockSizeVal != other.blockSizeVal) return false
            if (mToFingerprint != null) {
                if (other.mToFingerprint == null) return false
                if (!mToFingerprint.contentEquals(other.mToFingerprint)) return false
            } else if (other.mToFingerprint != null) return false
            if (mFromFingerprint != null) {
                if (other.mFromFingerprint == null) return false
                if (!mFromFingerprint.contentEquals(other.mFromFingerprint)) return false
            } else if (other.mFromFingerprint != null) return false
            if (extensionVal != other.extensionVal) return false
            if (hashlist != other.hashlist) return false
            if (sig != null) {
                if (other.sig == null) return false
                if (!sig.contentEquals(other.sig)) return false
            } else if (other.sig != null) return false
            if (filename != other.filename) return false
            if (mime != other.mime) return false
            if (endofstream != other.endofstream) return false

            return true
        }

        override fun hashCode(): Int {
            var result = toDisk.hashCode()
            result = 31 * result + (application?.contentHashCode() ?: 0)
            result = 31 * result + (sessionid ?: 0)
            result = 31 * result + (blockSizeVal ?: 0)
            result = 31 * result + (mToFingerprint?.contentHashCode() ?: 0)
            result = 31 * result + (mFromFingerprint?.contentHashCode() ?: 0)
            result = 31 * result + extensionVal.hashCode()
            result = 31 * result + (hashlist?.hashCode() ?: 0)
            result = 31 * result + (sig?.contentHashCode() ?: 0)
            result = 31 * result + (filename?.hashCode() ?: 0)
            result = 31 * result + (mime?.hashCode() ?: 0)
            result = 31 * result + endofstream.hashCode()
            return result
        }

        /**
         * Instantiates a new Builder.
         */
        init {
            sessionid = -1
            blockSizeVal = -1
            mime = "application/octet-stream"
        }
    }

    companion object {
        /**
         * New builder builder.
         *
         * @return the builder
         */
        @kotlin.jvm.JvmStatic
        fun newBuilder(): Builder {
            return Builder()
        }

        class Parser : ScatterSerializable.Companion.Parser<BlockData, BlockHeaderPacket>(BlockData.parser())
        fun parser(): Parser {
            return Parser()
        }
    }
}